File: M3104/tomcat-etud/webapps/gpsapp/WEB-INF/src/db/JourneyDAO.java
import db.SqliteConnection;

import java.sql.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;

public class JourneyDAO implements DAO<Journey>{
    private static JourneyDAO inst;
    private SqliteConnection connection = SqliteConnection.getInstance();

    public static JourneyDAO getInstance() {
        if (JourneyDAO.inst == null) {
            JourneyDAO.inst = new JourneyDAO();
        }
        return JourneyDAO.inst;
    }

    public int insert(Journey journey) {
        int newId = -1;
        Connection con = null;

        String query = "INSERT INTO Journey"
                + "(description) VALUES"
                + "(?)"
                + ";";

        try {
            con = connection.getConnection();
            PreparedStatement state = con.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
            state.setString(1, journey.getDescription());
            state.executeUpdate();

            ResultSet rs = state.getGeneratedKeys();
            if (rs.next()) {
                newId = rs.getInt(1);
                journey.setId(newId);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }

        return  newId;
    }

    public boolean update(Journey journey) {
        boolean ret = false;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "UPDATE Journey SET description = ?"
                    + ", distance = ?"
                    + " WHERE id = ?";
            PreparedStatement state = con.prepareStatement(query);

            state.setString(1, journey.getDescription());
            state.setFloat(2, journey.getDistance());
            state.setInt(3, journey.getId());

            state.executeUpdate();
            ret = true;
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    public boolean delete(Journey journey) {
        boolean ret = false;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "DELETE FROM journey "
                    + " WHERE id = ?"
                    + ";";
            PreparedStatement state = con.prepareStatement(query);
            state.setInt(1,  journey.getId());
            state.executeUpdate();
            ret = true;
        }
        catch (SQLException e) {
            e.printStackTrace();
        }
        finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return ret;
    }

    public Journey find(int id) {
        Journey journey = null;
        Connection con = null;
        try {
            con = connection.getConnection();
            String query = "SELECT * FROM Journey WHERE id = ?";
            PreparedStatement state = con.prepareStatement(query);
            state.setInt(1, id);
            ResultSet res = state.executeQuery();
            journey = new Journey(res.getString("description"));
            journey.setId(res.getInt("id"));
            journey.setDate_creation(new Date(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").parse(res.getString("date_creation")).getTime()));
            journey.setDistance(res.getFloat("distance"));
        }
        catch (SQLException e) {
            e.printStackTrace();
        } catch (ParseException e) {
            e.printStackTrace();
        } finally {
            try {
                con.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return journey;
    }

}

File: M3104/PHP/my_gps_web_app/model/JourneyDAO.php
<?php
require_once('SqliteConnection.php');
require_once("Journey.php");

class JourneyDAO {
    private static $dao;

    private function __construct() {
    }

    public final static function getInstance() {
        if(!isset(self::$dao)) {
            self::$dao= new JourneyDAO();
        }
        return self::$dao;
    }

    public final function find($id) {
        $dbc = SqliteConnection::getInstance()->getConnection();

        $stmt = $dbc->prepare("SELECT * FROM Journey WHERE id = :journey_id");
        $stmt->bindValue(':journey_id', (int) $id, PDO::PARAM_INT);
        $stmt->execute();
        $result = $stmt->fetch(PDO::FETCH_ASSOC);

        $journey = new Journey();
        $journey->setId($result["id"]);
        $journey->setDescription($result["description"]);
        $journey->setDateCreation($result["date_creation"]);
        $journey->setDistance($result["distance"]);
        return $journey;
    }

    public final function findAll(){
        $dbc = SqliteConnection::getInstance()->getConnection();
        $query = "SELECT * FROM Journey;";
        $stmt = $dbc->query($query);
        $results = $stmt->fetchALL(PDO::FETCH_CLASS, 'Journey');
        return $results;
    }

    public final function insert(Journey $journey){
        if($journey instanceof Journey){
            $dbc = SqliteConnection::getInstance()->getConnection();
            $stmt = $dbc->prepare("INSERT INTO Journey(description) values (:d)");
            $stmt->bindValue(':d', $journey->getDescription(), PDO::PARAM_STR);
            $stmt->execute();
            return true;
        }
    }

    public function delete(Journey $journey) {
        if($journey instanceof Journey) {
            $dbc = SqliteConnection::getInstance()->getConnection();
            $stmt = $dbc->prepare("DELETE FROM Journey WHERE id = :journey_id");
            $stmt->bindValue(':journey_id', $journey->getId(), PDO::PARAM_INT);
            $stmt->execute();
            return true;
        }
    }

    public function update(Journey $journey) {
        if($journey instanceof Journey) {
            $dbc = SqliteConnection::getInstance()->getConnection();
            $stmt = $dbc->prepare("UPDATE Journey SET description = :description, distance = :distance WHERE id = :id");
            $stmt->bindValue(':description', $journey->getDescription(), PDO::PARAM_STR);
            $stmt->bindValue(':distance', $journey->getDistance(), PDO::PARAM_STR);
            $stmt->bindValue(':id', $journey->getId(), PDO::PARAM_INT);
            $stmt->execute();
            return true;
        }
    }
}
?>
